src/include/transaction/deferred_action_manager.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+#include <queue>
+#include <utility>
+#include <vector>
+#include "transaction/timestamp_manager.h"
+#include "transaction/transaction_defs.h"
+
+namespace terrier::transaction {
+/**
+ * The deferred action manager tracks deferred actions and provides a function to process them
+ */
+class DeferredActionManager {
+ public:
+  /**
+   * Constructs a new DeferredActionManager
+   * @param timestamp_manager source of timestamps in the system
+   */
+  explicit DeferredActionManager(TimestampManager *timestamp_manager) : timestamp_manager_(timestamp_manager) {}
+
+  /**
+   * Adds the action to a buffered list of deferred actions.  This action will
+   * be triggered no sooner than when the epoch (timestamp of oldest running
+   * transaction) is more recent than the time this function was called.
+   * @param a functional implementation of the action that is deferred. @see DeferredAction
+   */
+  timestamp_t RegisterDeferredAction(const DeferredAction &a) {
+    timestamp_t result = timestamp_manager_->CurrentTime();
+    {
+      common::SpinLatch::ScopedSpinLatch guard(&deferred_actions_latch_);
+      new_deferred_actions_.emplace(result, a);
+    }
+    return result;
+  }
+
+  /**
+   * Adds the action to a buffered list of deferred actions.  This action will
+   * be triggered no sooner than when the epoch (timestamp of oldest running
+   * transaction) is more recent than the time this function was called.
+   * @param a functional implementation of the action that is deferred
+   */
+  timestamp_t RegisterDeferredAction(const std::function<void()> &a) {
+    // TODO(Tianyu): Will this be a performance problem? Hopefully C++ is smart enough
+    // to optimize out this call...
+    return RegisterDeferredAction([=](timestamp_t /*unused*/) { a(); });
+  }
+
+  /**
+   * Clear the queue and apply as many actions as possible
+   * @return numbers of deferred actions processed
+   */
+  uint32_t Process() {
+    uint32_t processed = 0;
+    // Check out a timestamp from the transaction manager to determine the progress of
+    // running transactions in the system.
+    timestamp_t oldest_txn = timestamp_manager_->OldestTransactionStartTime();
+    processed += ClearBacklog(oldest_txn);
+    processed += ProcessNewActions(oldest_txn);
+    return processed;
+  }
+
+ private:
+  TimestampManager *timestamp_manager_;
+  // TODO(Tianyu): We might want to change this data structure to be more specialized than std::queue
+  // Both queue should stay sorted in timestamp
```

Comment:
The queues are not guaranteed to be monotonically increasing on time stamp.  See note on register method.

Reply:
Yup. Leftover comment from previous version. Removed.